src/tools/remote-test-server/src/main.rs

```diff
@@ -43,18 +44,24 @@ static TEST: AtomicUsize = ATOMIC_USIZE_INIT;
 
 fn main() {
     println!("starting test server");
-    let listener = t!(TcpListener::bind("10.0.2.15:12345"));
+    let (listener, work) = if cfg!(target_os = "android") {
+        (t!(TcpListener::bind("0.0.0.0:12345")), "/data/tmp/work")
+    } else {
+        (t!(TcpListener::bind("10.0.2.15:12345")), "/tmp/work")
+    };
     println!("listening!");
 
-    let work = Path::new("/tmp/work");
+    let work = Path::new(work);
     t!(fs::create_dir_all(work));
 
     let lock = Arc::new(Mutex::new(()));
 
     for socket in listener.incoming() {
         let mut socket = t!(socket);
         let mut buf = [0; 4];
-        t!(socket.read_exact(&mut buf));
+        if socket.read_exact(&mut buf).is_err() {
```

Comment:
I've found that the way emulators treat network connections is super flaky. The server was crashing on startup on Android for a very weird reason, which I was later able to debug fully. The behavior I saw was:

* The client would need to push 4 dylibs initially to the server
* All 4 pushes finished *immediately*
* Oddly enough, only the last one would actually get pushed

Everything was "fixed" when I forced the client to keep the TCP connection alive until the server sent an "ack" back acknowledging it received the whole file. I have no idea why the data wasn't actually sent from the client to the server when it was all reported as successfully being sent on the client.

Given all that behavior what I think was happening was:

* The client created a TCP connection to the server (this goes through the emulation layer)
* The client successfully sent all data
* The client closed the TCP connection
* The server accepted the connection
* Somehow the emulator decided to not send all the data on the TCP connection and reported EOF

That EOF in turn previously caused a panic here. It ended up being much easier to debug if I ignored errors here and the server got farther. In general this seemed like a good idea because of how flaky the emulated networks seemed to be...